### PR TITLE
 Add a `joint_dof_mask` argument to `newton.ik.IKSolver`.

### PR DESCRIPTION
**Pull request by lyd405121**
## Description

  - Add a `joint_dof_mask` argument to `newton.ik.IKSolver`.
  - Allow selected joint DOFs to remain fixed while the LM optimizer updates the
  remaining DOFs.
  - Apply the mask consistently to analytic, autodiff, and mixed Jacobians.
  - Validate the mask dtype, shape, device, optimizer, and sampler
  configuration.
  - Add tests covering mask behavior and invalid configurations.

https://github.com/user-attachments/assets/6c634297-35da-436e-b3ea-c35925ccb6bb

  This is useful for humanoid control and other IK tasks where only a subset of
  the model's joint DOFs should be optimized, without rebuilding the model or
  modifying the IK objectives.

  ## Checklist

  - [x] New or existing tests cover these changes
  - [x] The documentation is up to date with these changes
  - [x] `CHANGELOG.md` has been updated (if user-facing change)

  ## Test plan

  ```bash
  uv run --extra dev -m newton.tests -k test_joint_dof_mask

  ## New feature / API change

  joint_dof_mask is a model-wide boolean array with one entry per joint DOF. A
  False entry keeps the corresponding DOF fixed, while a True entry allows it to
  be optimized.

  The mask currently supports the LM optimizer with sampling disabled.

  import warp as wp

  from newton import ik

  joint_dof_mask = wp.array(
      [False, True, True],
      dtype=wp.bool,
      device=model.device,
  )

  solver = ik.IKSolver(
      model,
      n_problems=1,
      objectives=objectives,
      optimizer=ik.IKOptimizer.LM,
      sampler=ik.IKSampler.NONE,
      joint_dof_mask=joint_dof_mask,
  )


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added an optional joint DOF mask to the inverse kinematics solver.
  * Selected joint degrees of freedom can now remain fixed during Levenberg–Marquardt optimization.

* **Bug Fixes**
  * Masking is now applied to Jacobian sensitivity so masked DOFs do not affect the LM linear solve.

* **Tests**
  * Added coverage for masked joint behavior across Jacobian modes and for rejecting invalid mask configurations.

* **Documentation**
  * Documented the new option in the unreleased changelog.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

### CODE REVIEW

**Comment by adenzler-nvidia** [comment]
Thank you very much - this looks like a very useful contribution. Deferring to @dylanturpin for an in-depth review as he authored most of this code.

File: newton/_src/sim/ik/ik_solver.py [hunk 1]
```diff
@@ -217,6 +217,10 @@ class IKSolver:
         lambda_min: Minimum LM damping value.
         lambda_max: Maximum LM damping value.
         rho_min: Minimum LM acceptance ratio.
+        joint_dof_mask: Optional model-wide mask with one value per joint DOF.
```
Comment (dylanturpin) [inline newton/_src/sim/ik/ik_solver.py]: Worth stating both polarities + the shape here — "mask" is ambiguous (numpy's convention is the opposite). Also note it's dof-space indexing, so a free joint has 6 entries not 7.

File: newton/_src/sim/ik/ik_lm_optimizer.py [hunk 2]
```diff
@@ -142,6 +154,7 @@ def __init__(
         rho_min: float = 1e-3,
         *,
         problem_idx: wp.array[wp.int32] | None = None,
+        joint_dof_mask: wp.array[wp.bool] | None = None,
```
Comment (dylanturpin) [inline newton/_src/sim/ik/ik_lm_optimizer.py:190]: IKOptimizerLM is public API too but skips all the validation — a wrong-length mask here silently reads past the end of the array. I'd move the dtype/shape/device checks into this constructor and let IKSolver inherit them.

**Review by dylanturpin — COMMENTED** [review]
Thanks — this looks great and is definitely a feature req I've heard before. Mechanism is the right one: zeroing J columns before the damped solve gives exactly-zero updates for masked dofs + doesn't contaminate the step for the free ones (better than clamping dq after the solve).

This all lgtm, just one thing to address (+ some nits):

**free/ball/distance joints**: the mask acts in twist space, but the integrator couples a quaternion joint's dofs — so partially masking one doesn't pin its coordinates. Masking a free joint's 3 linear dofs to fix base position still lets the base drift (~0.1–0.35 over 50 iters when I tried it), which is exactly the floating-base humanoid case. Fully-masked quaternion joints keep their pose (up to quat renormalization roundoff), so I'd document the twist-space semantics, reject partial masks over free/ball/distance in validation, and add a free-joint test.

To make this easy I put ready-to-cherry-pick commits with the above + the nits + a rebase onto current main here: https://github.com/dylanturpin/newton-collab/tree/pr3488-suggestions (`214430c2` is the substantive one — cherry-picks cleanly onto your branch; mask + IK convergence tests pass on cpu and cuda). Take them, adapt, or I can push them onto this PR — whatever's easiest.

File: newton/tests/test_ik.py [hunk 3]
```diff
@@ -373,6 +373,76 @@ def test_convergence_mixed_d6(test, device):
...
+            [position_objective],
+            jacobian_mode=mode,
+            joint_dof_mask=wp.array([False, True], dtype=wp.bool, device=device),
+        )
+
+        solver.step(joint_q, joint_q, iterations=40)
+
+        result = joint_q.numpy()[0]
+        test.assertAlmostEqual(float(result[0]), 0.4, places=6)
+        test.assertGreater(abs(float(result[1])), 1.0e-3)
```
Comment (dylanturpin) [inline newton/tests/test_ik.py]: Forgot to assert it actually converges — this only checks the free dof moved. Also the masked assert can be exact (`assertEqual`), the masked delta is exactly 0 by construction.

File: newton/tests/test_ik.py [hunk 4]
```diff
@@ -373,6 +373,76 @@ def test_convergence_mixed_d6(test, device):
     _convergence_test_d6(test, device, ik.IKJacobianType.MIXED)
 
 
+def test_joint_dof_mask(test, device, mode: ik.IKJacobianType):
```
Comment (dylanturpin) [inline newton/tests/test_ik.py:376]: Coverage nits: a free-joint case (the motivating one), n_problems >= 2 so the batch dim of the new kernel gets exercised, + all-True mask == no mask.

File: newton/_src/sim/ik/ik_lm_optimizer.py [hunk 5]
```diff
@@ -80,6 +80,16 @@ def _update_lm_state(
         lambda_values[row] = wp.clamp(new_lambda, lambda_min, lambda_max)
 
 
+@wp.kernel
+def _zero_masked_jacobian_columns(
```
Comment (dylanturpin) [inline newton/_src/sim/ik/ik_lm_optimizer.py]: nit: name reads inverted — it zeroes the columns where the mask is False. `_zero_fixed_dof_jacobian_columns`?

File: CHANGELOG.md [hunk 6]
```diff
@@ -24,6 +24,7 @@
 - Add USD import support for `NewtonJointAPI` (`newton:armature`, `newton:damping`, `newton:friction`, `newton:velocityLimit`, `newton:limitStiffness`, `newton:limitDamping`). Attributes broadcast uniformly to every DOF on the joint. If per DOF variance is required, recommendation is to break apart into 1-DOF (i.e. revolute & prismatic) joints instead.
 - Add support for pt2 neural-network checkpoints (saved via `torch.export.save`) in `ControllerNeuralMLP` and `ControllerNeuralLSTM`.
 - Add `SensorContact.position_matrix` alongside `force_matrix`, reporting per-counterpart world-frame contact positions (force-weighted average of contact midpoints).
+- Add `joint_dof_mask` to `newton.ik.IKSolver` to keep selected joint DOFs fixed during LM optimization.
```
Comment (dylanturpin) [inline CHANGELOG.md]: nit: after rebasing this needs to move to the new Unreleased section (1.4.0 shipped since your base).

File: newton/_src/sim/ik/ik_solver.py (hunk 1 above)
Comment (lyd405121) [reply newton/_src/sim/ik/ik_solver.py]: - Sorry，I forgot it
- I only test revolute and prismatic joint

**Comment by lyd405121** [comment]
> Thanks — this looks great and is definitely a feature req I've heard before. Mechanism is the right one: zeroing J columns before the damped solve gives exactly-zero updates for masked dofs + doesn't contaminate the step for the free ones (better than clamping dq after the solve).
> 
> This all lgtm, just one thing to address (+ some nits):
> 
> **free/ball/distance joints**: the mask acts in twist space, but the integrator couples a quaternion joint's dofs — so partially masking one doesn't pin its coordinates. Masking a free joint's 3 linear dofs to fix base position still lets the base drift (~0.1–0.35 over 50 iters when I tried it), which is exactly the floating-base humanoid case. Fully-masked quaternion joints keep their pose (up to quat renormalization roundoff), so I'd document the twist-space semantics, reject partial masks over free/ball/distance in validation, and add a free-joint test.
> 
> To make this easy I put ready-to-cherry-pick commits with the above + the nits + a rebase onto current main here: https://github.com/dylanturpin/newton-collab/tree/pr3488-suggestions (`214430c2` is the substantive one — cherry-picks cleanly onto your branch; mask + IK convergence tests pass on cpu and cuda). Take them, adapt, or I can push them onto this PR — whatever's easiest.

- I will pick your code into my branch, and merge lastest newton code
- All of your advice are precise and right

**Review by eric-heiden — APPROVED** [review]
